docs/motivation.md

```diff
@@ -0,0 +1,553 @@
+type COrD = C | D
+```
+This describes the union of the two types.
+
+If we assume `C` and `D` above are classes, we can furthermore write pattern
+matches to distinguish between the two types dynamically.
+
+Since in Verona there is no subtyping between classes, we are able to check that
+a pattern match is exhaustive, and furthermore matching on concrete types will
+allow us to do implicit type filtering over match cases. E.g. imagine a case
+match like this
+```verona
+// C, D, E are classes
+
+f(x : C | D | E, g : (D | E) -> S) : S {
+    match x with {
+        (y : C) => { // y : C
+            ...
+        }
+        _ => { // x : D | E
+            g x
+        }
+    }
+}
+```
+We are able to implicitly filter the type of `x` in the default case.
+
+In a language like ML, the corresponding code would look like
+```ocaml
+type t = C of c
+       | D of d
+       | E of e
+
+
+f (x : t) (g : t -> s) =
+    match x with
+        C y => ...
+        _   => g x
+```
+Note that `g` cannot have a more refined type than `t -> s`.
+
+#### Type discrimination
+In ML we might want to construct a type such as
+```ocaml
+type s = A of int
+         B of int
+```
+A corresponding Verona type might look like
+```verona
+type S = I32 | I32
+```
+but it is now impossible to discriminate between the two cases. This is however
+easily fixed by defining `S` instead as
+```verona
+class A { var i : I32 }
+class B { var i : I32 }
+
+type S = A | B
+```
+
+#### Explicit nullability
+We can easily encode nullability with a class `None`
+```verona
+type A
+class None {}
+
+f(x : A | None) : S {
+    match x with {
+        (y: A) => { // y : A
+            ...
+        }
+        _ => { // x : None
+            ...
+        }
+    }
+}
+```
+
+### Subtyping and inheritance
+Verona is an object-oriented system with classes and traits. When it comes to
+subtyping, classes are treated nominally, and traits are treated structurally.
+Furthermore there is no subtyping between classes. This means a class
+can be a subtype of a trait, but never a subtype of a class other than itself.
+For example, given the definitions
+```verona
+class C {
+    equal(self: C, other: C): Bool {...}
+}
+
+class D {
+    equal(self: D, other: D): Bool {...}
+}
+
+type Eq = {
+    equal(self: Self, other: Self) : Bool
+}
+
+```
+both `C` and `D` are subtypes of `Eq`, but neither `C` or `D` are subtypes of
+the other.
+
+#### Inheritance
+There are two forms of inheritance in
+Verona. One between classes and one between class and trait.
+
+##### Class inheritance
+Inheritance between classes works as you would expect in a language like
+Java, with the exception that it does not imply subtyping. As an example,
+consider
+```verona
+class C {
+    equal(self: C, other: C): Bool {...}
+    f(...) : ... {...}
+}
+
+// here we reuse functionality of C in D, specifically the method f, but
+// we do not get a subtyping relation D <: C
+class D : C {
+    equal(self: D, other: D): Bool {...}
```

Comment:
What would have happened if `D` hadn't redefined `equal` like this?  Would it have picked up an `equal : (C,C) -> Bool` that then could be called with `::` as in `D::equal(c1,c2)` but not via `.`, since `D </: C`?

Reply:
Yes. Caveat: mentioning your own type could also be treated as `Self`, so that D got `equal: (Self, Self)->Bool`, that needs discussion (it may be that such a trick leads to too many type errors in the body of the function).

More generally: I think we should favor the term "code reuse" instead of "inheritance".